contracts/src/FunctionGateway.sol

```diff
@@ -2,246 +2,270 @@
+    bool public isCallback;
 
-    function initialize(uint256 _scalar, address _feeVault, address _timelock, address _guardian)
-        external
-        initializer
-    {
-        scalar = _scalar;
-        feeVault = _feeVault;
+    /// @dev Initializes the contract.
+    /// @param _timelock The address of the timelock contract.
+    /// @param _guardian The address of the guardian.
+    function initialize(
+        address _timelock,
+        address _guardian
+    ) external initializer {
+        isCallback = false;
         __TimelockedUpgradeable_init(_timelock, _guardian);
     }
 
-    function request(bytes32 _functionId, bytes memory _input, bytes4 _callbackSelector, bytes memory _context)
-        external
-        payable
-        returns (bytes32)
-    {
-        return request(_functionId, _input, _callbackSelector, _context, DEFAULT_GAS_LIMIT, tx.origin);
-    }
-
-    /// @dev Requests for a proof to be generated by the marketplace.
-    /// @param _functionId The id of the proof to be generated.
-    /// @param _input The input to the proof.
-    /// @param _context The context of the runtime.
+    /// @dev Creates a onchain request for a proof. The output and proof is fulfilled asynchronously
+    ///      by the provided callback.
+    /// @param _functionId The function identifier.
+    /// @param _input The function input.
+    /// @param _context The function context.
     /// @param _callbackSelector The selector of the callback function.
-    function request(
+    /// @param _callbackGasLimit The gas limit for the callback function.
+    function zkRequest(
         bytes32 _functionId,
         bytes memory _input,
-        bytes4 _callbackSelector,
         bytes memory _context,
-        uint256 _gasLimit,
-        address _refundAccount
-    ) public payable returns (bytes32) {
+        bytes4 _callbackSelector,
+        uint32 _callbackGasLimit
+    ) external payable returns (bytes32) {
+        // Compute the callback hash uniquely associated with this request.
         bytes32 inputHash = sha256(_input);
-        bytes32 contextHash = keccak256(_context);
-        FunctionRequest memory r = FunctionRequest({
-            functionId: _functionId,
-            inputHash: inputHash,
-            outputHash: bytes32(0),
-            contextHash: contextHash,
-            callbackAddress: msg.sender,
-            callbackSelector: _callbackSelector,
-            proofFulfilled: false,
-            callbackFulfilled: false
-        });
-
-        uint256 feeAmount = _handlePayment(_gasLimit, _refundAccount, msg.sender, msg.value);
-
-        bytes32 requestId = keccak256(abi.encode(nonce, r));
-        requests[requestId] = r;
-
-        emit ProofRequested(nonce, _functionId, requestId, _input, _context, _gasLimit, feeAmount);
-        nonce++;
-        return requestId;
-    }
+        bytes32 contextHash = sha256(_context);
```

Comment:
nit: here we should do `keccack` since it's cheaper (also in our circuit once we have an optimized keccack implementation, we should maybe do that for the input as well)?